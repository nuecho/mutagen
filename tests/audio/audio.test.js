
const DEPENDENCIES_FILE_PATH = getResourcePath("audio/dependencies.csv");
const INVALID_FILE_PATH = getResourcePath("audio/invalid.csv");
const INVALID_PERSONALITY_PATH = getResourcePath("audio/invalid-personality.csv");
const ISO_LATIN_FILE_PATH = getResourcePath("audio/import-iso-latin.csv");
const MINIMALIST_FILE_PATH = getResourcePath("audio/import-minimalist.csv");
const MISSING_AUDIO_FILES_PATH = getResourcePath("audio/missing-audio-files.csv");
const VALID_FILE_PATH = getResourcePath("audio/import.csv");

const OUTPUT_PREFIX = "/tmp/mutagen/test";

beforeAll(() => {
  const { code } = mutagen(`--env gax audio import ${DEPENDENCIES_FILE_PATH}`);
  expect(code).toBe(0);
})

test("[mutagen audio --help] should print audio usage", () => {
  assertMutagenResult(`audio --help`, "audio usage", 0);
});

///////////
// Audio Export

test("[mutagen audio export --help] should print export usage", () => {
  assertMutagenResult(`--env gax audio export --help`, "export usage", 0);
});

test("[mutagen audio export] should print usage when output file is not specified", () => {
  assertMutagenResult(`--env gax audio export`, "export usage", 1);
});

test("[mutagen audio export] should export messages into a csv file", () => {
  const outputFile = `${OUTPUT_PREFIX}/csv-only/output.csv`;
  assertMutagenResult(`--env gax audio export ${outputFile}`, "export done", 0);

  const csvFile = readCsv(outputFile, "export_test");
  expect(csvFile).toMatchSnapshot("csv file");
});

test("[mutagen audio export --personality-ids=x] should export messages into a csv file only for selected personalities", () => {
  const outputFile = `${OUTPUT_PREFIX}/personalities/output.csv`;
  assertMutagenResult(`--env gax audio export --personality-ids=11 ${outputFile}`, "export done", 0);

  const csvFileContent = readCsv(outputFile, "export_test");
  expect(csvFileContent).toMatchSnapshot("csv file");
});

test("[mutagen audio export --with-audios] should export audio files", () => {
  const outputDirectory = `${OUTPUT_PREFIX}/with-audios`;
  assertMutagenResult(`--env gax audio export --with-audios ${outputDirectory}/output.csv`, "export done", 0);

  const downloadedFiles = exec(`find ${outputDirectory} -type f -name *export_test* | sort`);
  expect(downloadedFiles.output).toMatchSnapshot("downloaded files");
});

///////////
// Audio Import

test(`[mutagen audio import --help] should print import usage`, () => {
  assertMutagenResult(`--env gax audio import --help`, "import usage", 0);
});

test(`[mutagen audio import] should print usage when file to import is not specified`, () => {
  assertMutagenResult(`--env gax audio import`, "import usage", 1);
});

test(`[mutagen audio import] should fail when file to import does not exist`, () => {
  assertMutagenResult(`--env gax audio import /tmp/foo`, "no such file or directory", 1);
});

test(`[mutagen audio import] should fail when file to import is not a valid CSV`, () => {
  assertMutagenResult(`--env gax audio import ${INVALID_FILE_PATH}`, "invalid csv", 1);
});

test(`[mutagen audio import] should fail when some audio files are missing`, () => {
  assertMutagenResult(`--env gax audio import ${MISSING_AUDIO_FILES_PATH}`, "missing audio files", 1);
});

test(`[mutagen audio import] should fail when the csv file references invalid personality`, () => {
  assertMutagenResult(`--env gax audio import ${INVALID_PERSONALITY_PATH}`, "missing personality", 1);
});

test(`[mutagen audio import] should import messages and audio files`, () => {
  testImportWithAudios(VALID_FILE_PATH, "import_test");
});

test(`[mutagen audio import] message type and description should be optional`, () => {
  testImportWithAudios(MINIMALIST_FILE_PATH, "import_minimalist_test");
});

test(`[mutagen audio import] should allow setting input file encoding with --encoding`, () => {
  testImportWithAudios(ISO_LATIN_FILE_PATH, "import_iso_latin_test", "--encoding=iso-8859-1");
});

test(`[mutagen audio import] should fail when the specified encoding is invalid`, () => {
  assertMutagenResult(`--env gax audio import --encoding=POTATO-8 ${VALID_FILE_PATH}`, "invalid encoding message", 1);
});

test(`[mutagen audio import] should validate duplicated messages`, () => {
  assertMutagenResult(`--env gax audio import ${VALID_FILE_PATH}`, "duplicated messages", 1);
});

///////////
// Utilities

function testImportWithAudios(importFilePath, messagesKey, additionalArgument = "") {
  assertMutagenResult(`--env gax audio import ${additionalArgument} ${importFilePath}`, "mutagen import output", 0);

  const outputDirectory = `${OUTPUT_PREFIX}/${messagesKey}`;
  assertMutagenResult(`--env gax audio export --with-audios ${outputDirectory}/output.csv`, "mutagen export output", 0);

  const findResult = exec(`find ${outputDirectory} -type f -name *${messagesKey}* | sort`);
  expect(findResult.output).toMatchSnapshot("export of the imported files");

  const catResult = readCsv(`${outputDirectory}/output.csv`, messagesKey);
  expect(catResult).toMatchSnapshot("re-export of the csv file");
}

const readCsv = (file, keyword) => {
  // About the sed: we are replacing the messageArId columns by <removed-for-testing-purposes> since we can't be sure
  // what the ID will be when comparing with the snapshot.
  return exec(`cat ${file} | egrep '(messageArId)|(${keyword})' | sed -r "s/,[0-9]{4},/,<removed-for-testing-purposes>,/g"`).output
}