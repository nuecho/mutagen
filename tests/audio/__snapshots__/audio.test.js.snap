// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[mutagen audio --help] should print audio usage: audio usage 1`] = `
Object {
  "stderr": "",
  "stdout": "Usage: audio [-?]
GAX ARM tool
  -?, -h, --help              Show this help message.
Commands:
  import  Import audio files in ARM
  export  Export audio files from ARM
",
}
`;

exports[`[mutagen audio export --help] should print export usage: export usage 1`] = `
Object {
  "stderr": "",
  "stdout": "Usage: export [-?] [--with-audios] [--gax-api-path=<gaxApiPath>]
              [--personality-ids=personality-id[,personality-id]...]...
              outputFile
Export audio files from ARM
      outputFile              Output audio description file.
      --gax-api-path=<gaxApiPath>
                              GAX API path. Defaults to '/gax/api'.
      --personality-ids=personality-id[,personality-id]...
                              Comma separated list of which personalities'
                                audios to export (10,11,12,...).
      --with-audios           Download messages' audio files.
  -?, -h, --help              Show this help message.
",
}
`;

exports[`[mutagen audio export --personality-ids=x] should export messages into a csv file only for selected personalities: csv file 1`] = `
Object {
  "stderr": "",
  "stdout": "name,type,description,messageArId,tenantId,11
\\"export_test music name 1\\",MUSIC,\\"this is the description for music name 1\\",<removed-for-testing-purposes>,1,\\"11/export_test music name 1.wav\\"
\\"export_test message name 1\\",ANNOUNCEMENT,\\"this is the description for message name 1\\",<removed-for-testing-purposes>,1,\\"11/export_test message name 1.wav\\"
\\"export_test music name 2\\",MUSIC,\\"this is the description for music name 2\\",<removed-for-testing-purposes>,1,
\\"export_test message name 2\\",ANNOUNCEMENT,\\"this is the description for message name 2\\",<removed-for-testing-purposes>,1,
\\"export_test music name 3\\",MUSIC,\\"this is the description for music name 3\\",<removed-for-testing-purposes>,1,\\"11/export_test music name 3.wav\\"
",
}
`;

exports[`[mutagen audio export --personality-ids=x] should export messages into a csv file only for selected personalities: export done 1`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`[mutagen audio export --with-audios] should export audio files: downloaded files 1`] = `
Object {
  "stderr": "",
  "stdout": "/tmp/mutagen/test/with-audios/10/export_test message name 1.wav
/tmp/mutagen/test/with-audios/10/export_test message name 2.wav
/tmp/mutagen/test/with-audios/10/export_test music name 3.wav
/tmp/mutagen/test/with-audios/11/export_test message name 1.wav
/tmp/mutagen/test/with-audios/11/export_test music name 1.wav
/tmp/mutagen/test/with-audios/11/export_test music name 3.wav
",
}
`;

exports[`[mutagen audio export --with-audios] should export audio files: export done 1`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`[mutagen audio export] should export messages into a csv file: csv file 1`] = `
Object {
  "stderr": "",
  "stdout": "name,type,description,messageArId,tenantId,10,11
\\"export_test music name 1\\",MUSIC,\\"this is the description for music name 1\\",<removed-for-testing-purposes>,1,,\\"11/export_test music name 1.wav\\"
\\"export_test message name 1\\",ANNOUNCEMENT,\\"this is the description for message name 1\\",<removed-for-testing-purposes>,1,\\"10/export_test message name 1.wav\\",\\"11/export_test message name 1.wav\\"
\\"export_test music name 2\\",MUSIC,\\"this is the description for music name 2\\",<removed-for-testing-purposes>,1,,
\\"export_test message name 2\\",ANNOUNCEMENT,\\"this is the description for message name 2\\",<removed-for-testing-purposes>,1,\\"10/export_test message name 2.wav\\",
\\"export_test music name 3\\",MUSIC,\\"this is the description for music name 3\\",<removed-for-testing-purposes>,1,\\"10/export_test music name 3.wav\\",\\"11/export_test music name 3.wav\\"
",
}
`;

exports[`[mutagen audio export] should export messages into a csv file: export done 1`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`[mutagen audio export] should fail when GAX API path is invalid: invalid GAX API path 1`] = `
Object {
  "stderr": "Failed to log in to GAX.
",
  "stdout": "",
}
`;

exports[`[mutagen audio export] should print usage when output file is not specified: export usage 1`] = `
Object {
  "stderr": "",
  "stdout": "Missing required parameter: outputFile
Usage: export [-?] [--with-audios] [--gax-api-path=<gaxApiPath>]
              [--personality-ids=personality-id[,personality-id]...]...
              outputFile
Export audio files from ARM
      outputFile              Output audio description file.
      --gax-api-path=<gaxApiPath>
                              GAX API path. Defaults to '/gax/api'.
      --personality-ids=personality-id[,personality-id]...
                              Comma separated list of which personalities'
                                audios to export (10,11,12,...).
      --with-audios           Download messages' audio files.
  -?, -h, --help              Show this help message.
",
}
`;

exports[`[mutagen audio import --help] should print import usage: import usage 1`] = `
Object {
  "stderr": "",
  "stdout": "Usage: import [-?] [--encoding=<encoding>] [--gax-api-path=<gaxApiPath>]
              inputFile
Import audio files in ARM
      inputFile               Input audio description file.
      --encoding=<encoding>   Encoding of the input file. Defaults to UTF-8.
      --gax-api-path=<gaxApiPath>
                              GAX API path. Defaults to '/gax/api'.
  -?, -h, --help              Show this help message.
",
}
`;

exports[`[mutagen audio import] message type and description should be optional: export of the imported files 1`] = `
Object {
  "stderr": "",
  "stdout": "/tmp/mutagen/test/import_minimalist_test/10/import_minimalist_test message name 1.wav
/tmp/mutagen/test/import_minimalist_test/10/import_minimalist_test message name 2.wav
",
}
`;

exports[`[mutagen audio import] message type and description should be optional: mutagen export output 1`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`[mutagen audio import] message type and description should be optional: mutagen import output 1`] = `
Object {
  "stderr": "",
  "stdout": "Imported 2 messages.
",
}
`;

exports[`[mutagen audio import] message type and description should be optional: re-export of the csv file 1`] = `
Object {
  "stderr": "",
  "stdout": "name,type,description,messageArId,tenantId,10,11
\\"import_minimalist_test message name 1\\",ANNOUNCEMENT,,<removed-for-testing-purposes>,1,\\"10/import_minimalist_test message name 1.wav\\",
\\"import_minimalist_test message name 2\\",ANNOUNCEMENT,,<removed-for-testing-purposes>,1,\\"10/import_minimalist_test message name 2.wav\\",
",
}
`;

exports[`[mutagen audio import] should allow setting input file encoding with --encoding: export of the imported files 1`] = `
Object {
  "stderr": "",
  "stdout": "/tmp/mutagen/test/import_iso_latin_test/10/import_iso_latin_test.wav
",
}
`;

exports[`[mutagen audio import] should allow setting input file encoding with --encoding: mutagen export output 1`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`[mutagen audio import] should allow setting input file encoding with --encoding: mutagen import output 1`] = `
Object {
  "stderr": "",
  "stdout": "Imported 1 message.
",
}
`;

exports[`[mutagen audio import] should allow setting input file encoding with --encoding: re-export of the csv file 1`] = `
Object {
  "stderr": "",
  "stdout": "name,type,description,messageArId,tenantId,10,11
import_iso_latin_test,ANNOUNCEMENT,\\"pâtatë poèl\\",<removed-for-testing-purposes>,1,\\"10/import_iso_latin_test.wav\\",
",
}
`;

exports[`[mutagen audio import] should fail when GAX API path is invalid: invalid GAX API path 1`] = `
Object {
  "stderr": "Failed to log in to GAX.
",
  "stdout": "",
}
`;

exports[`[mutagen audio import] should fail when file to import does not exist: no such file or directory 1`] = `
Object {
  "stderr": "/tmp/foo (No such file or directory)
",
  "stdout": "",
}
`;

exports[`[mutagen audio import] should fail when file to import is not a valid CSV: invalid csv 1`] = `
Object {
  "stderr": "Invalid input file.
",
  "stdout": "",
}
`;

exports[`[mutagen audio import] should fail when some audio files are missing: missing audio files 1`] = `
Object {
  "stderr": "Audio import failed: the following audio files don't exist
/usr/src/app/tests/resources/audio/11/MISSING AUDIO FILE.wav
",
  "stdout": "",
}
`;

exports[`[mutagen audio import] should fail when the csv file references invalid personality: missing personality 1`] = `
Object {
  "stderr": "Audio import failed: the following personalities don't exist
12
",
  "stdout": "",
}
`;

exports[`[mutagen audio import] should fail when the specified encoding is invalid: invalid encoding message 1`] = `
Object {
  "stderr": "",
  "stdout": "Could not convert 'POTATO-8' to Charset for option '--encoding': java.nio.charset.UnsupportedCharsetException: POTATO-8
Usage: import [-?] [--encoding=<encoding>] [--gax-api-path=<gaxApiPath>]
              inputFile
Import audio files in ARM
      inputFile               Input audio description file.
      --encoding=<encoding>   Encoding of the input file. Defaults to UTF-8.
      --gax-api-path=<gaxApiPath>
                              GAX API path. Defaults to '/gax/api'.
  -?, -h, --help              Show this help message.
",
}
`;

exports[`[mutagen audio import] should import messages and audio files: export of the imported files 1`] = `
Object {
  "stderr": "",
  "stdout": "/tmp/mutagen/test/import_test/10/import_test message name 1.wav
/tmp/mutagen/test/import_test/10/import_test message name 2.wav
/tmp/mutagen/test/import_test/10/import_test music name 3.wav
/tmp/mutagen/test/import_test/11/import_test message name 1.wav
/tmp/mutagen/test/import_test/11/import_test music name 1.wav
/tmp/mutagen/test/import_test/11/import_test music name 3.wav
",
}
`;

exports[`[mutagen audio import] should import messages and audio files: mutagen export output 1`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`[mutagen audio import] should import messages and audio files: mutagen import output 1`] = `
Object {
  "stderr": "",
  "stdout": "Imported 5 messages.
",
}
`;

exports[`[mutagen audio import] should import messages and audio files: re-export of the csv file 1`] = `
Object {
  "stderr": "",
  "stdout": "name,type,description,messageArId,tenantId,10,11
\\"import_test music name 1\\",MUSIC,\\"this is the description for music name 1\\",<removed-for-testing-purposes>,1,,\\"11/import_test music name 1.wav\\"
\\"import_test message name 1\\",ANNOUNCEMENT,\\"this is the description for message name 1\\",<removed-for-testing-purposes>,1,\\"10/import_test message name 1.wav\\",\\"11/import_test message name 1.wav\\"
\\"import_test music name 2\\",MUSIC,\\"this is the description for music name 2\\",<removed-for-testing-purposes>,1,,
\\"import_test message name 2\\",ANNOUNCEMENT,\\"this is the description for message name 2\\",<removed-for-testing-purposes>,1,\\"10/import_test message name 2.wav\\",
\\"import_test music name 3\\",MUSIC,\\"this is the description for music name 3\\",<removed-for-testing-purposes>,1,\\"10/import_test music name 3.wav\\",\\"11/import_test music name 3.wav\\"
",
}
`;

exports[`[mutagen audio import] should print usage when file to import is not specified: import usage 1`] = `
Object {
  "stderr": "",
  "stdout": "Missing required parameter: inputFile
Usage: import [-?] [--encoding=<encoding>] [--gax-api-path=<gaxApiPath>]
              inputFile
Import audio files in ARM
      inputFile               Input audio description file.
      --encoding=<encoding>   Encoding of the input file. Defaults to UTF-8.
      --gax-api-path=<gaxApiPath>
                              GAX API path. Defaults to '/gax/api'.
  -?, -h, --help              Show this help message.
",
}
`;

exports[`[mutagen audio import] should validate duplicated messages: duplicated messages 1`] = `
Object {
  "stderr": "Audio import failed: the following messages' names already exist on the gax server
import_test music name 1
import_test message name 1
import_test music name 2
import_test message name 2
import_test music name 3
",
  "stdout": "",
}
`;
